File: internal/wire/agent.go
package wire

// Local-agent REST + WS wire types. Mirrors src/wire/agent.ts.
// Reference source: ~/Documents/dev/meta-me-local-agent.

// ─── Threads + messages ────────────────────────────────────────────────

type AgentThread struct {
	ID            string  `json:"id"`
	Title         string  `json:"title"`
	ProjectID     *string `json:"project_id"`
	ModelProvider *string `json:"model_provider"`
	ModelID       *string `json:"model_id"`
	UpdatedAt     int64   `json:"updated_at"`
	MsgCount      *int    `json:"msg_count,omitempty"`
}

type AgentThreadsListResp struct {
	Threads []AgentThread `json:"threads"`
}

type AgentMessage struct {
	ID        string `json:"id"`
	Role      string `json:"role"`
	Content   string `json:"content"`
	CreatedAt int64  `json:"created_at"`
}

type AgentThreadMessagesResp struct {
	Messages []AgentMessage `json:"messages"`
}

type AgentMessageSearchMatch struct {
	AgentMessage
	ThreadID string `json:"thread_id"`
	Title    string `json:"title"`
}

type AgentMessageSearchResp struct {
	Matches []AgentMessageSearchMatch `json:"matches"`
}

// ─── Projects ──────────────────────────────────────────────────────────

type AgentProject struct {
	ID           string  `json:"id"`
	Label        string  `json:"label"`
	RootPath     string  `json:"root_path"`
	LastOpenedAt *int64  `json:"last_opened_at"`
	ThreadCount  *int    `json:"thread_count,omitempty"`
}

type AgentProjectsListResp struct {
	Projects []AgentProject `json:"projects"`
}

// ─── Models ────────────────────────────────────────────────────────────

type AgentModel struct {
	Provider string   `json:"provider"`
	ID       string   `json:"id"`
	Label    string   `json:"label"`
	Input    []string `json:"input"`
}

type AgentModelsListResp struct {
	Models []AgentModel `json:"models"`
}

// ─── Health ────────────────────────────────────────────────────────────

type AgentHealthResp struct {
	OK         bool   `json:"ok"`
	TS         int64  `json:"ts"`
	Version    string `json:"version"`
	Uptime     int    `json:"uptime"`
	Sockets    int    `json:"sockets"`
	Sessions   int    `json:"sessions"`
	InstallID  *string `json:"installId"`
}

File: src/wire/agent.ts
/**
 * Local-agent REST + WS wire types — request/response shapes for every
 * endpoint mm-cli calls on `http://localhost:3142` (or the tailnet URL
 * of a `--node` target).
 *
 * Naming: `Agent<Resource><Verb>` for REST; `AgentWs<EventType>` for WS.
 *
 * Reference for the agent source: ~/Documents/dev/meta-me-local-agent.
 */

// ─── Threads + messages ────────────────────────────────────────────────

export type AgentThread = {
	id: string;
	title: string;
	project_id: string | null;
	model_provider: string | null;
	model_id: string | null;
	updated_at: number;
	msg_count?: number;
};

export type AgentThreadsListResp = {
	threads: AgentThread[];
};

export type AgentMessage = {
	id: string;
	role: 'user' | 'assistant' | 'system' | 'tool';
	content: string;
	created_at: number;
};

export type AgentThreadMessagesResp = {
	messages: AgentMessage[];
};

export type AgentMessageSearchMatch = AgentMessage & {
	thread_id: string;
	title: string;
};

export type AgentMessageSearchResp = {
	matches: AgentMessageSearchMatch[];
};

// ─── Projects ──────────────────────────────────────────────────────────

export type AgentProject = {
	id: string;
	label: string;
	root_path: string;
	last_opened_at: number | null;
	thread_count?: number;
};

export type AgentProjectsListResp = {
	projects: AgentProject[];
};

// ─── Models ────────────────────────────────────────────────────────────

export type AgentModel = {
	provider: string;
	id: string;
	label: string;
	input: string[];
};

export type AgentModelsListResp = {
	models: AgentModel[];
};

// ─── Health ────────────────────────────────────────────────────────────

export type AgentHealthResp = {
	ok: boolean;
	ts: number;
	version: string;
	uptime: number;
	sockets: number;
	sessions: number;
	installId: string | null;
	tcc: {
		ok: boolean;
		platform: string;
		checkedAt: number;
		failedPaths: string[];
		recentDenialAt?: number;
	};
};

// ─── WebSocket protocol — outbound (CLI → agent) ───────────────────────

export type AgentWsSend = {
	type: 'send';
	threadId: string;
	content: string;
	provider?: string;
	modelId?: string;
	projectId?: string;
};

export type AgentWsResume = {
	type: 'resume';
	threadId: string;
	cursor: number;
};

export type AgentWsPing = { type: 'ping' };

export type AgentWsOutbound = AgentWsSend | AgentWsResume | AgentWsPing;

// ─── WebSocket protocol — inbound (agent → CLI) ────────────────────────

export type AgentWsDelta = {
	type: 'delta';
	threadId: string;
	cursor: number;
	text: string;
};

export type AgentWsThinkingDelta = {
	type: 'thinking_delta';
	threadId: string;
	cursor: number;
	text: string;
};

export type AgentWsToolStart = {
	type: 'tool_start';
	threadId: string;
	cursor: number;
	toolName: string;
	args?: unknown;
};

export type AgentWsToolEnd = {
	type: 'tool_end';
	threadId: string;
	cursor: number;
	toolName: string;
	result?: unknown;
};

export type AgentWsStatus = {
	type: 'status';
	threadId: string;
	cursor: number;
	message: string;
};

export type AgentWsDone = {
	type: 'done';
	threadId: string;
	cursor: number;
	fullText?: string;
	fullThinking?: string;
};

export type AgentWsError = {
	type: 'error';
	threadId: string;
	cursor: number;
	message: string;
};

export type AgentWsResumeEmpty = {
	type: 'resume_empty';
	threadId: string;
};

export type AgentWsInbound =
	| AgentWsDelta
	| AgentWsThinkingDelta
	| AgentWsToolStart
	| AgentWsToolEnd
	| AgentWsStatus
	| AgentWsDone
	| AgentWsError
	| AgentWsResumeEmpty;
